From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Tue, 9 Jul 2024 19:12:41 +0300
Subject: [PATCH] Injecting custom mob effects

Displaying effects to the client is done on the plugin's side.
This simply integrates custom effects into the general API.

Note that effects are constructed & added into the registry after plugins' onLoad took place, so
interacting with effects registry before onEnable won't have the new values in it.
This is done on purpuse and is not a bug.

diff --git a/src/main/java/me/sosedik/kiterino/util/IKiterinoBootstrapMobEffectInjector.java b/src/main/java/me/sosedik/kiterino/util/IKiterinoBootstrapMobEffectInjector.java
new file mode 100644
index 0000000000000000000000000000000000000000..76013d0dabc933553c18bd49f1f3673679659a0b
--- /dev/null
+++ b/src/main/java/me/sosedik/kiterino/util/IKiterinoBootstrapMobEffectInjector.java
@@ -0,0 +1,21 @@
+package me.sosedik.kiterino.util;
+
+import me.sosedik.kiterino.world.effect.KiterinoNMSMobEffect;
+import org.bukkit.NamespacedKey;
+import org.bukkit.potion.PotionEffectType;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.Contract;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.function.Supplier;
+
+@ApiStatus.Internal
+interface IKiterinoBootstrapMobEffectInjector {
+
+	void injectMobEffects(@NotNull Class<?> mobEffectsClass, @NotNull String namespace);
+
+	@Contract("_, _ -> null")
+	@Nullable PotionEffectType injectMobEffect(@Nullable NamespacedKey key, @NotNull Supplier<? extends KiterinoNMSMobEffect> mobEffectSupplier);
+
+}
diff --git a/src/main/java/me/sosedik/kiterino/util/KiterinoBootstrapMobEffectInjector.java b/src/main/java/me/sosedik/kiterino/util/KiterinoBootstrapMobEffectInjector.java
new file mode 100644
index 0000000000000000000000000000000000000000..3ae6f3f22260a345e3be0edba724ea0552f6d74e
--- /dev/null
+++ b/src/main/java/me/sosedik/kiterino/util/KiterinoBootstrapMobEffectInjector.java
@@ -0,0 +1,83 @@
+package me.sosedik.kiterino.util;
+
+import io.papermc.paper.plugin.bootstrap.BootstrapContext;
+import me.sosedik.kiterino.world.effect.KiterinoNMSMobEffect;
+import org.bukkit.NamespacedKey;
+import org.bukkit.potion.PotionEffectType;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.Locale;
+import java.util.function.Supplier;
+
+public class KiterinoBootstrapMobEffectInjector {
+
+	private KiterinoBootstrapMobEffectInjector() {
+		throw new IllegalStateException("Utility class");
+	}
+
+	static IKiterinoBootstrapMobEffectInjector injector;
+
+	/**
+	 * Inject materials in the class using the plugin's name as namespace.
+	 * Keys will be fetched from Material field names.
+	 *
+	 * <p><b>Can be used only during bootstrap phase.
+	 *
+	 * @param mobEffectsClass class with public static final Material instances
+	 * @param context bootstrap context
+	 * @throws IllegalStateException if not in bootstrap phase
+	 */
+	public static void injectMobEffects(@NotNull Class<?> mobEffectsClass, @NotNull BootstrapContext context) {
+		injectMobEffects(mobEffectsClass, context.getConfiguration().getName().toLowerCase(Locale.ROOT).replace("-", "_"));
+	}
+
+	/**
+	 * Inject materials in the class using the provided namespace.
+	 * Keys will be fetched from Material field names.
+	 *
+	 * <p><b>Can be used only during bootstrap phase.
+	 *
+	 * @param mobEffectsClass class with public static final Material instances
+	 * @param namespace namespace for Materials
+	 * @throws IllegalStateException if not in bootstrap phase
+	 */
+	public static void injectMobEffects(@NotNull Class<?> mobEffectsClass, @NotNull String namespace) {
+		if (injector == null) throw new IllegalStateException("Can't inject mob effects outside bootstrap phase!");
+		injector.injectMobEffects(mobEffectsClass, namespace);
+	}
+
+	/**
+	 * Injects custom mob effect.
+	 * The key will be calculated from Plugin's name as namespace
+	 * and mob effect's name as a key.
+	 *
+	 * <p><b>Can be used only during bootstrap phase.
+	 *
+	 * @param options nms mob effect
+	 * @return injected mob effect
+	 * @throws IllegalStateException if not in bootstrap phase
+	 */
+	public static @NotNull PotionEffectType injectMobEffect(@NotNull Supplier<? extends KiterinoNMSMobEffect> options) {
+		return injectMobEffect(null, options);
+	}
+
+	/**
+	 * Injects custom mob effect.
+	 * If the key is null, it will be calculated from Plugin's provided namespace
+	 * and mob effect field's name as a key.
+	 *
+	 * <p><b>Can be used only during bootstrap phase.
+	 *
+	 * @param key material key
+	 * @param options nms mob effect
+	 * @return injected mob effect
+	 * @throws IllegalStateException if not in bootstrap phase
+	 */
+	public static @NotNull PotionEffectType injectMobEffect(@Nullable NamespacedKey key, @NotNull Supplier<? extends KiterinoNMSMobEffect> options) {
+		if (injector == null) throw new IllegalStateException("Can't inject mob effects outside bootstrap phase!");
+		// Yes, null is expected; the field will be set later via reflection
+		return injector.injectMobEffect(key, options);
+	}
+
+}
diff --git a/src/main/java/me/sosedik/kiterino/world/effect/KiterinoMobEffect.java b/src/main/java/me/sosedik/kiterino/world/effect/KiterinoMobEffect.java
new file mode 100644
index 0000000000000000000000000000000000000000..bed9bf017f3d945dbe01e9c546385436ed76f38a
--- /dev/null
+++ b/src/main/java/me/sosedik/kiterino/world/effect/KiterinoMobEffect.java
@@ -0,0 +1,118 @@
+package me.sosedik.kiterino.world.effect;
+
+import org.bukkit.damage.DamageSource;
+import org.bukkit.entity.Entity;
+import org.bukkit.entity.LivingEntity;
+import org.bukkit.potion.PotionEffectTypeCategory;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.jetbrains.annotations.Range;
+
+/**
+ * Wrapper for simple Kiterino mob effects
+ */
+public interface KiterinoMobEffect extends KiterinoNMSMobEffect {
+
+	/**
+	 * Gets the effect's category
+	 *
+	 * @return effect's category
+	 */
+	@NotNull PotionEffectTypeCategory getCategory();
+
+	/**
+	 * Gets the effect's color value
+	 *
+	 * @return effect's color value
+	 */
+	int getColor();
+
+	/**
+	 * Gets the effect's blend duration ticks
+	 *
+	 * @return effect's blend duration ticks
+	 */
+	default int getBlendDurationTicks() {
+		return 0;
+	}
+
+	/**
+	 * Called when the effect is ticked.
+	 * If {@code false} is returned, the effect will be removed from the entity.
+	 *
+	 * @param entity entity that has the effect
+	 * @param amplifier effect amplifier
+	 * @return whether the effect was ticked
+	 */
+	default boolean applyEffectTick(@NotNull LivingEntity entity, int amplifier) {
+		return true;
+	}
+
+	/**
+	 * Called when the effect is applied if this effect is {@link #isInstantaneous()}
+	 *
+	 * @param source effect source (e.g. thrown potion)
+	 * @param attacker attacker entity (e.g. thrown potion's owner)
+	 * @param target entity the effect is applied to
+	 * @param amplifier effect amplifier
+	 * @param proximity intensity relative to maximum effect; 0.0: not affected; 1.0: fully hit by potion effects
+	 */
+	default void applyInstantaneousEffect(@Nullable Entity source, @Nullable Entity attacker, @NotNull LivingEntity target, int amplifier, @Range(from = 0L, to = 1L) double proximity) {
+		this.applyEffectTick(target, amplifier);
+	}
+
+	/**
+	 * Checks whether the effect's tick method ({@link #applyEffectTick(LivingEntity, int)}) should be fired this tick
+	 *
+	 * @param duration effect duration
+	 * @param amplifier effect amplifier
+	 * @return whether to call tick method
+	 */
+	default boolean shouldApplyEffectTickThisTick(int duration, int amplifier) {
+		return false;
+	}
+
+	/**
+	 * Called when the effect was added or updated to the entity
+	 *
+	 * @param entity entity
+	 * @param amplifier effect amplifier
+	 */
+	default void onEffectStarted(@NotNull LivingEntity entity, int amplifier) {}
+
+	/**
+	 * Called when the effect is being freshly added to the entity
+	 *
+	 * @param entity entity
+	 * @param amplifier effect amplifier
+	 */
+	default void onEffectAdded(@NotNull LivingEntity entity, int amplifier) {}
+
+	/**
+	 * Called when the entity having this effect dies
+	 *
+	 * @param entity entity
+	 * @param amplifier effect amplifier
+	 */
+	default void onMobKilled(@NotNull LivingEntity entity, int amplifier) {}
+
+	/**
+	 * Called when the entity having this effect is hurt
+	 *
+	 * @param entity entity
+	 * @param amplifier effect amplifier
+	 * @param source damage source
+	 * @param amount damage amount
+	 */
+	default void onMobHurt(@NotNull LivingEntity entity, int amplifier, @NotNull DamageSource source, float amount) {}
+
+	/**
+	 * Checks whether this effect is instantaneous
+	 *
+	 * @return whether this effect is instantaneous
+	 */
+	default boolean isInstantaneous() {
+		return false;
+	}
+
+}
diff --git a/src/main/java/me/sosedik/kiterino/world/effect/KiterinoNMSMobEffect.java b/src/main/java/me/sosedik/kiterino/world/effect/KiterinoNMSMobEffect.java
new file mode 100644
index 0000000000000000000000000000000000000000..f90a8fc0797b9f231da51e28aee128c81e9e2b51
--- /dev/null
+++ b/src/main/java/me/sosedik/kiterino/world/effect/KiterinoNMSMobEffect.java
@@ -0,0 +1,9 @@
+package me.sosedik.kiterino.world.effect;
+
+/**
+ * Represents Kiterino mob effect
+ * (used for injecting custom mob effects into registry).
+ */
+public interface KiterinoNMSMobEffect {
+
+}
